### PR TITLE
travis: Run thread sanitizer on unit tests

### PR DESCRIPTION
**Pull request by maflcko**
On unit tests only for now. Disabled for the gui unit tests and all functional tests.

### CODE REVIEW

**Comment by sipa** [comment]
This looks related: https://groups.google.com/forum/#!topic/thread-sanitizer/nmhRglhlohI

**Comment by practicalswift** [comment]
Strong concept ACK

We really should test under all sanitizers in Travis – thanks for taking a step in that direction! :-)

**Comment by laanwj** [comment]
utACK

**Review by practicalswift — CHANGES_REQUESTED** [review]
Excellent work @MarcoFalke!

utACK modulo suggested changes

Would be really nice to also have the functional tests running under TSan :-)

File: .travis/test_03_before_install.sh [hunk 1]
```diff
@@ -8,7 +8,8 @@ export LC_ALL=C.UTF-8
 
 PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
 # Add llvm-symbolizer directory to PATH. Needed to get symbolized stack traces from the sanitizers.
-PATH=$PATH:/usr/lib/llvm-6.0/bin/
+# shellcheck disable=SC2125
+PATH=$PATH:/usr/lib/llvm-*/bin/
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply .travis/test_03_before_install.sh]: If there is only one folder that matches the wildcard, it should be properly expanded?

However, I just went and reverted that hunk.

File: .travis.yml [hunk 2]
```diff
@@ -94,23 +94,24 @@ jobs:
...
-# x86_64 Linux (xenial, no depends, only system libs)
+# x86_64 Linux (xenial, no depends, only system libs, sanitizers: thread (TSAN))
     - stage: test
       env: >-
         HOST=x86_64-unknown-linux-gnu
         DOCKER_NAME_TAG=ubuntu:16.04
-        PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools libssl-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
+        PACKAGES="clang llvm python3-zmq qtbase5-dev qttools5-dev-tools libssl-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
         NO_DEPENDS=1
+        RUN_FUNCTIONAL_TESTS=false
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply .travis.yml]: They also need to be identified, which is the tricky part and imo better be suited for another pull request. (See my WIP in the suppressions file)

File: .travis/test_04_install.sh [hunk 3]
```diff
@@ -7,8 +7,9 @@
 export LC_ALL=C.UTF-8
 
 travis_retry docker pull "$DOCKER_NAME_TAG"
+export TSAN_OPTIONS="suppressions=${TRAVIS_BUILD_DIR}/test/sanitizer_suppressions/tsan"
 export UBSAN_OPTIONS="suppressions=${TRAVIS_BUILD_DIR}/contrib/sanitizers-ubsan.suppressions:print_stacktrace=1:halt_on_error=1"
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply .travis/test_04_install.sh]: Thanks, done.

File: .travis/test_03_before_install.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (practicalswift) [reply .travis/test_03_before_install.sh]: @MarcoFalke No, the preceding `$PATH:` part prevents expansion:

```
$ mkdir testing
$ ls -d testin*
testing
$ TESTING=foo/:testin*
$ echo $TESTING
foo/:testin*
$ TESTING=testin*
$ echo $TESTING
testing
```

:-)

**Pull request merged**